pydantic/_internal/_generate_schema.py

```diff
@@ -292,16 +292,24 @@ def _typed_dict_schema(self, typed_dict_cls: Any) -> core_schema.TypedDictSchema
         """
         Generate schema for a TypedDict.
         """
+        type_hints = _typing_extra.get_type_hints(typed_dict_cls, include_extras=True)
+        required_keys: typing.FrozenSet[str]
         try:
-            required_keys: typing.FrozenSet[str] = typed_dict_cls.__required_keys__
+            required_keys = typed_dict_cls.__required_keys__
         except AttributeError:
-            raise TypeError('Please use `typing_extensions.TypedDict` instead of `typing.TypedDict`.')
+            # __required_keys__ is only available on Python >= 3.11
+            # if it is not present then _all_ keys are either jointly required (total=True)
+            # or optional (total=False)
```

Comment:
do you know in pre-3.11 if you could achieve more nuanced logic as a function of base classes? I'm thinking the answer is yes. I defer to you/Samuel what we should do before __required_keys__ was available, but it is probably worth noting that technically this might not perfectly match the defined logic.

What I mean is:
```python
class MyTotalTypedDict(TypedDict):
    x: int

class MyLessTotalTypedDict(MyTotalTypedDict):
    y: int
```
Even in versions of python without `__required_keys__` I think `x` might be required and `y` not required.

Reply:
Let me take a look into the implementation of __required_keys__ in 3.11. Maybe we can just copy that

Reply:
But off the top of my head yes, you're right I think we actually need to traverse all the base classes to get the right logic. I expect implementation won't be that bad, but the test might be super ugly if we want to cover all the edge cases.